lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp

```diff
@@ -0,0 +1,267 @@
+#include "lazy_tensor_core/csrc/ops/constant_pad_nd.h"
+#include "lazy_tensor_core/csrc/ops/device_data.h"
+#include "lazy_tensor_core/csrc/ops/expand.h"
+#include "lazy_tensor_core/csrc/ops/ltc_ops.h"
+#include "lazy_tensor_core/csrc/ops/permute.h"
+#include "lazy_tensor_core/csrc/ops/scalar.h"
+#include "lazy_tensor_core/csrc/ops/unsqueeze.h"
+#include "lazy_tensor_core/csrc/ops/view.h"
+#include "lazy_tensor_core/csrc/ts_backend/ts_computation_client.h"
+#include "lazy_tensor_core/csrc/ts_backend/ts_lowering_context.h"
+#include "lazy_tensors/permutation_util.h"
+
+namespace torch_lazy_tensors {
+namespace compiler {
+
+class TSNodeLowering : public NodeLowering {
+ public:
+  TSNodeLowering(const std::string& name, ts_backend::TSLoweringContext* loctx)
+      : NodeLowering(loctx),
+        function_(loctx ? std::make_shared<torch::jit::GraphFunction>(
+                              name, loctx->graph(), nullptr)
+                        : nullptr) {}
+
+  bool Lower(const ir::Node* node) override {
+    TSOpVector ops = LowerToTS(node);
+    if (ops.empty()) {
+      return false;
+    }
+    LTC_CHECK_EQ(node->num_outputs(), ops.size());
+    for (size_t i = 0; i < ops.size(); ++i) {
+      loctx()->AssignOutputOp(ir::Output(node, i), ops[i]);
+    }
+    return true;
+  }
+
+  lazy_tensors::Shape Infer(const ir::Node* node) override {
+    if (node->op().op == at::aten::expand) {
+      auto expand =
+          ir::NodeCast<ir::ops::Expand>(node, ir::OpKind(at::aten::expand));
+      const ir::Output& argument = node->operand(0);
+      return lazy_tensors::Shape(argument.shape().element_type(),
+                                 expand->size());
+    }
+    if (node->op().op == at::aten::permute) {
+      auto permute =
+          ir::NodeCast<ir::ops::Permute>(node, ir::OpKind(at::aten::permute));
+      const ir::Output& argument = node->operand(0);
+      return ir::ops::Permute::MakePermuteShape(argument.shape(),
+                                                permute->dims());
+    }
+    if (node->op().op == at::aten::constant_pad_nd) {
+      auto constant_pad_nd = ir::NodeCast<ir::ops::ConstantPadNd>(
+          node, ir::OpKind(at::aten::constant_pad_nd));
+      const ir::Output& argument = node->operand(0);
+      const lazy_tensors::Shape& argument_shape = argument.shape();
+      const auto argument_dimensions = argument_shape.dimensions();
+      const auto& pad = constant_pad_nd->pad();
+      LTC_CHECK_EQ(argument_dimensions.size() * 2, pad.size());
+      std::vector<lazy_tensors::int64> padded_dimensions(
+          argument_dimensions.begin(), argument_dimensions.end());
+      size_t i = 0;
+      for (auto rit = pad.rbegin(); rit != pad.rend(); rit += 2, ++i) {
+        padded_dimensions[i] += (*rit + *(rit + 1));
+      }
+      return lazy_tensors::Shape(argument_shape.element_type(),
+                                 padded_dimensions);
+    }
+    LTC_LOG(FATAL) << *node << "Not implemented yet.";
```

Comment:
i'm not clear on which cases Infer even gets called.  where do I look to understand that?